neo/Network/LocalNode.cs

```diff
@@ -204,22 +205,49 @@ public static void AllowHashes(IEnumerable<UInt256> hashes)
         private void Blockchain_PersistCompleted(object sender, Block block)
         {
             Transaction[] remain;
+            var millisSinceLastBlock = TimeSpan.FromTicks(DateTimeOffset.UtcNow.Ticks)
+                .Subtract(TimeSpan.FromTicks(LastBlockReceived.Ticks)).TotalMilliseconds;
+
             lock (mem_pool)
             {
+                // Remove the transactions that made it into the block
                 foreach (Transaction tx in block.Transactions)
-                {
                     mem_pool.Remove(tx.Hash);
-                }
                 if (mem_pool.Count == 0) return;
 
                 remain = mem_pool.Values.ToArray();
                 mem_pool.Clear();
+                
```

Comment:
I don't understand the problem in this point, would you explain me this block a little more please?

Reply:
Sure. So the problem is that anything in the mem_pool after a block is persisted needs to be re-verified, because it could be invalid now that we are at a new block height. So before we just move it all into temp_pool to be verified; however, suppose there are some hooks that should run after a block that need to see transactions that were pending in the mem_pool to do something consensus related. If we don't re-verify the mem_pool immediately then it will be empty during those hooks (the Persist_Unlocked hook). Since we only really care that the transactions are re-verified in the case when we are at the tip of the chain, which will mean that the last block generally seen greater than 10 seconds ago, we can reverify immediately with the mem_pool lock held. So any transactions that didn't get confirmed in the last block that were in the mem_pool that are still valid, but didn't make it into the last block appear to stay in the mem_pool by anyone looking at the mem_pool before and after the last block at the tip of the chain was accepted.